README.md
<h1 align="center" xmlns="http://www.w3.org/1999/html">
  <br>
   <img src="https://user-images.githubusercontent.com/64373963/158168111-a773b83e-71be-47dd-a43b-8e12b0c33c8d.png" width=860 />
  <br>
</h1>


<p align="center">
  <img src="https://img.shields.io/pub/v/fluid_action_card?color=%23004088&logo=flutter">
  <img src="https://img.shields.io/github/license/Shantanu66/Fluid-Action-Card?color=%23A100FF">
  <img src="https://img.shields.io/badge/Animation%20-Package-yellow">
  <img src="https://img.shields.io/github/repo-size/Shantanu66/Fluid-Action-Card?color=orange">
  <img src="https://img.shields.io/github/stars/Shantanu66/Fluid-Action-Card?style=social">
</p>




# Fluid action card Package

A Flutter package consisting of pre-animated cards(containers) with fluid animation for freely adding user customized cards to the app with heavily customizable options for the cards adding up to an incredible UI experience.You can set the amount of cards for your screen to display without writing code for creating additonal cards and also writing your own animation mechanism.You can also add multiple texts and images in the card and also customize its color which is in gradient form with various attributes making it a complete package for modern frontend development.


# Preview 
![Screenshot 2022-03-16 135641](https://user-images.githubusercontent.com/64373963/158547901-a374f7d2-5b32-4a35-8999-fe1e73052364.png)

# Demo
<p align="center">
<img src="https://user-images.githubusercontent.com/64373963/158566922-07af4787-a543-4ccb-866a-8883b10d533d.gif">
</p>

# Installation
You should have [Flutter](https://flutter.dev/) installed in your system.<br/>
Now in your terminal <br/>
Run this command:<br/>

    $ flutter pub add fluid_action_card
    
  



# Example and Usage
<p align="center">
<img src="https://user-images.githubusercontent.com/64373963/153234878-f23e4f19-1ef5-4e55-8bd8-c52249b59459.png" width="310" height="520">
</p>


```dart

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        title: 'Fluid Action card display',
        debugShowCheckedModeBanner: false,
        home: Scaffold(
          body: FluidActionCard(
            color1: Colors.pinkAccent,
            color2: Colors.black45,
            backgroundcolor: Colors.grey[900]!,
            borderRadius1: BorderRadius.circular(20),
            borderRadius2: BorderRadius.circular(20),
            height: 400.0,
            width: 240.0,
            cardCount: 6,
            Position: 250.0,
            shadow: BoxShadow(
              color: Colors.black12,
              blurRadius: 10.0,
              spreadRadius: 0.2,
              offset: Offset(0, 3),
            ),
            ontap: () {},
           ),
        )
      );
  }
}
```
<p align="center">
<img src="https://user-images.githubusercontent.com/64373963/153338603-e4072f98-1fcc-4bcb-a3b0-f2f0c4a979bf.png" width="310" height="460">

> To add images to the animated cards follow this procedure:
* First import the dependancy assets to pubspec.yaml

```dart
flutter:
  uses-material-design: true

  assets:
    - assets/images/
```

```dart
void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        title: 'Fluid Action card display',
        debugShowCheckedModeBanner: false,
        home: Scaffold(
          body: FluidActionCard(
            color1: Colors.greenAccent.shade400,
            color2: Colors.black12,
            backgroundcolor: Colors.grey[900]!,
            borderRadius1: BorderRadius.circular(20),
            borderRadius2: BorderRadius.circular(20),
            TextPosition_Top: 22.0,
            TextPosition_Down: 29.0,
            height: 400.0,
            width: 240.0,
            CardCount: 3,
            text1: Text(
              "Starts From",
              style: TextStyle(
                  color: Colors.white,
                  fontSize: 18.0,
                  fontStyle: FontStyle.italic),
            ),
            text2: Text(
              "300",
              style: TextStyle(
                color: Colors.white,
                fontSize: 20.0,
                fontWeight: FontWeight.w700,
                letterSpacing: 2.0,
              ),
            ),
            Position: 250.0,
            shadow: BoxShadow(
              color: Colors.black12,
              blurRadius: 10.0,
              spreadRadius: 0.2,
              offset: Offset(0, 3),
            ),
            assetimage: 'assets/images/Xbox.png',
            ontap: () {},
          ),
        ));
  }
}
```
# Support the Library
You can support the library by liking it on pub, staring in on Github and reporting any bugs you encounter.Feel Free to add to it and make it better by contributing to it


CHANGELOG.md
[1.0.0+1] - Release.

[1.0.0+2] 
 - Added image support
 - fixed bugs
[1.0.0+3]
 - improved the description
[1.0.0+4]
 - added example
[1.0.0+5]
 - improved the Readme

test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';


void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/UI/Ui Screen.dart
import 'dart:ffi';

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_plant_shop_ui/Assets/ImagesDesc.dart';

class SwipeBox extends StatefulWidget {
  final Color color1;

  final Color color2;
  final Color backgroundcolor;

  final BorderRadius borderRadius1;

  final BorderRadius borderRadius2;

  final double height;

  final double width;

  final AssetImage image1;

  final int BoxCount;

  final double Position;

  final BoxShadow shadow;
  final Function ontap;

  const SwipeBox(
      {Key key,
      @required this.color1,
      @required this.color2,
      @required this.borderRadius1,
      @required this.borderRadius2,
      @required this.height,
      @required this.width,
      @required this.image1,
      @required this.backgroundcolor,
      @required this.BoxCount,
      @required this.Position,
      this.shadow, this.ontap})
      : super(key: key);

  @override
  _SwipeBoxState createState() => _SwipeBoxState();
}

class _SwipeBoxState extends State<SwipeBox>
    with SingleTickerProviderStateMixin {
  TabController _tabController;
  PageController _pageController;
  int _selectedPage = 0;

  @override
  void initState() {
    super.initState();
    _tabController = TabController(initialIndex: 0, length: 5, vsync: this);
    _pageController = PageController(initialPage: 0, viewportFraction: 0.8);
  }

  _SwipeSelector(int index) {
    return AnimatedBuilder(
      animation: _pageController,
      builder: (BuildContext context, Widget widget) {
        double value = 1;
        if (_pageController.position.haveDimensions) {
          value = _pageController.page - index;
          value = (1 - (value.abs() * 0.4)).clamp(0.0, 1.0);
        }
        return Center(
          child: SizedBox(
            height: Curves.easeInOut.transform(value) * 500.0,
            width: Curves.easeInOut.transform(value) * 430.0,
            child: widget,
          ),
        );
      },
      child: GestureDetector(
        onTap: () {
          widget.ontap;
        },
      child: Stack(
        alignment: Alignment.center,
        children: <Widget>[
          ClipRRect(
            borderRadius: widget.borderRadius1,
            child: Container(
              decoration: BoxDecoration(
                gradient: LinearGradient(
                    begin: Alignment.topCenter,
                    end: Alignment.bottomCenter,
                    colors: [
                      widget.color1,
                      widget.color2,
                    ]),
                boxShadow: [widget.shadow],
                borderRadius: widget.borderRadius2,
              ),
              margin: EdgeInsets.fromLTRB(10.0, 10.0, 10.0, 30.0),
              child: Stack(
                children: <Widget>[
                  Center(
                    child: Hero(
                      tag: plants[index].imageUrl,
                      child: Image(
                        height: 270.0,
                        width: 290.0,
                        image: widget.image1,
                        fit: BoxFit.cover,
                      ),
                    ),
                  ),
                ],
              ),
            ),
          ),
        ],
      ),
    ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        backgroundColor: widget.backgroundcolor,
        body: AnnotatedRegion<SystemUiOverlayStyle>(
            value: SystemUiOverlayStyle.dark,
            child: ListView(
                padding: EdgeInsets.symmetric(vertical: widget.Position),
                children: <Widget>[
                  Container(
                    height: widget.height,
                    width: widget.width,
                    child: PageView.builder(
                      controller: _pageController,
                      onPageChanged: (int index) {
                        setState(() {
                          _selectedPage = index;
                        });
                      },
                      itemCount: widget.BoxCount,
                      itemBuilder: (BuildContext context, int index) {
                        return _SwipeSelector(index);
                      },
                    ),
                  ),
                ])));
  }
}


lib/Assets/ImagesDesc.dart
class Plant {
  final String imageUrl;
  final String name;
  final String category;
  final int price;
  final String size;
  final String description;

  Plant({
    this.imageUrl,
    this.name,
    this.category,
    this.price,
    this.size,
    this.description,
  });
}

final List<Plant> plants = [
  Plant(
    imageUrl: 'assets/images/plant0.png',
    name: 'Playstation 5',
    category: 'CONSOLE',
    price: 500,
    size: 'large',
    description:
        'he PlayStation 5 (PS5) is a home video game console developed by Sony Interactive Entertainment.'
            ' The PlayStation 5\'s main hardware features include a solid-state drive customized for high-speed data streaming to enable significant improvements in storage performance, an AMD GPU capable of 4K resolution display at up to 120 frames per second, hardware-accelerated ray tracing for realistic lighting and reflections, and the Tempest Engine allowing for hardware-accelerated 3D audio effects. Other features include the DualSense controller with haptic feedback and backward compatibility with most of the PlayStation 4 and PlayStation VR games.'
,
  ),
  Plant(
    imageUrl: 'assets/images/plant1.png',
    name: 'Xbox Series X/S',
    category: 'CONSOLE',
    price: 300,
    size: 'Large',
    description:
        'The Xbox Series X and the Xbox Series S (collectively, the Xbox Series X/S[b]) are home video game consoles developed by Microsoft. They were both released on November 10, 2020 as the fourth generation of the Xbox console family, succeeding the Xbox One family. Along with Sony\'s PlayStation 5, also released in November 2020, the Xbox Series X and Series S are part of the ninth generation of video game consoles'
    'The Xbox Series X has higher end hardware, and supports higher display resolutions (up to 8K resolution) along with higher frame rates and real-time ray tracing; it also has a high-speed solid-state drive to reduce loading times. The less expensive Xbox Series S uses the same CPU, but has a less powerful GPU, has less memory and internal storage, and lacks an optical drive. Both consoles are designed to support nearly all Xbox One games, controllers, and accessories, including those games from older Xbox consoles supported by Xbox One\'s backwards compatibility. ',
  ),
  Plant(
    imageUrl: 'assets/images/plant2.png',
    name: 'Nintendo Switch',
    category: 'CONSOLE',
    price: 300,
    size: 'small',
    description:
        'The Nintendo Switch[h] is a video game console developed by Nintendo and released worldwide in most regions on March 3, 2017. The console itself is a tablet that can either be docked for use as a home console or used as a portable device, making it a hybrid console'
      'As an eighth-generation console, the Nintendo Switch competes with Microsoft\'s Xbox One and Sonys PlayStation 4. Nearly three million console units were shipped in its first month, exceeding Nintendo\'s initial projection of two million and within a year of release achieved over 14 million units sold worldwide, outselling total lifetime sales of the Wii U. By the start of 2018, the Switch became the fastest-selling home or hybrid console in both Japan and the United States. As of June 2021, the Nintendo Switch and Nintendo Switch Lite have sold more than 89 million units worldwide. Switch sales have been strongly tied to sales of Nintendo\'s first-party titles, with six games—The Legend of Zelda: Breath of the Wild, Mario Kart 8 Deluxe, Super Mario Odyssey, Super Smash Bros. Ultimate, Pokémon Sword and Shield, and Animal Crossing: New Horizons—having sold over twenty million units each.'
  ),
];


lib/SwipeCard/SwipeCard.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

class SwipeCard extends StatefulWidget {
  //for setting the card's first gradient color
  final Color color1;
  //for setting the card's second gradient color
  final Color color2;
  //for setting the screen(container) background color
  final Color backgroundcolor;
  //for setting the card's borderradius and shape
  final BorderRadius borderRadius1;
  //for setting the card's borderradius and shape
  final BorderRadius borderRadius2;
  //for setting the card's height
  final double height;
  //for setting the card's width
  final double width;
  //for inserting image in the card
  final Image? image1;
  //for the number of cards the user wants
  final int BoxCount;
  //for setting the position of the card
  final double Position;
  //for setting the shadow of the card
  final BoxShadow? shadow;
  //for gesture detection of the card
  final Function? ontap;
  //for adding text to the card
  final Text? text1;
  //for adding text to the card
  final Text? text2;
  //for setting image height of the card
  final double? ImageHeight;
  //for setting image width of the card
  final double? ImageWidhth;
  //for setting text position on the card
  final double? TextPosition_Top;
  //for setting text position on the card
  final double? TextPosition_Down;
  //for setting different images on each card using tags
  final Hero? hero;

  const SwipeCard(
      {Key? key,
      required this.color1,
      required this.color2,
      required this.borderRadius1,
      required this.borderRadius2,
      required this.height,
      required this.width,
      this.image1,
      required this.backgroundcolor,
      required this.BoxCount,
      required this.Position,
      this.shadow,
      this.ontap,
      this.text1,
      this.text2,
      this.ImageHeight,
      this.ImageWidhth,
      this.TextPosition_Top,
      this.TextPosition_Down,
      this.hero})
      : super(key: key);

  @override
  _SwipeCardState createState() => _SwipeCardState();
}

class _SwipeCardState extends State<SwipeCard>
    with SingleTickerProviderStateMixin {
  TabController? _tabController;
  PageController? _pageController;
  int _selectedPage = 0;

  @override
  void initState() {
    super.initState();
    _tabController = TabController(initialIndex: 0, length: 5, vsync: this);
    _pageController = PageController(initialPage: 0, viewportFraction: 0.8);
  }

  _SwipeSelector(int index) {
    return AnimatedBuilder(
      animation: _pageController!,
      builder: (BuildContext context, Widget? widget) {
        double value = 1;
        if (_pageController!.position.haveDimensions) {
          value = (_pageController!.page! - index);
          value = (1 - (value.abs() * 0.4)).clamp(0.0, 1.0);
        }
        return Center(
          child: SizedBox(
            height: Curves.easeInOut.transform(value) * 500.0,
            width: Curves.easeInOut.transform(value) * 430.0,
            child: widget,
          ),
        );
      },
      child: GestureDetector(
        onTap: () {
          widget.ontap;
        },
        child: Stack(
          alignment: Alignment.center,
          children: <Widget>[
            ClipRRect(
              borderRadius: widget.borderRadius1,
              child: Container(
                decoration: BoxDecoration(
                  gradient: LinearGradient(
                      begin: Alignment.topCenter,
                      end: Alignment.bottomCenter,
                      colors: [
                        widget.color1,
                        widget.color2,
                      ]),
                  boxShadow: [widget.shadow!],
                  borderRadius: widget.borderRadius2,
                ),
                margin: EdgeInsets.fromLTRB(10.0, 10.0, 10.0, 30.0),
                child: Stack(
                  children: <Widget>[
                    Center(child: widget.hero),
                    Positioned(
                      top: widget.TextPosition_Top,
                      right: widget.TextPosition_Down,
                      child: Column(
                        crossAxisAlignment: CrossAxisAlignment.center,
                        children: <Widget>[
                          if (widget.text1 != null) widget.text1!,
                          if (widget.text2 != null) widget.text2!,
                        ],
                      ),
                    ),
                  ],
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        backgroundColor: widget.backgroundcolor,
        body: AnnotatedRegion<SystemUiOverlayStyle>(
            value: SystemUiOverlayStyle.dark,
            child: ListView(
                padding: EdgeInsets.symmetric(vertical: widget.Position),
                children: <Widget>[
                  Container(
                    height: widget.height,
                    width: widget.width,
                    child: PageView.builder(
                      controller: _pageController,
                      onPageChanged: (int index) {
                        setState(() {
                          _selectedPage = index;
                        });
                      },
                      itemCount: widget.BoxCount,
                      itemBuilder: (BuildContext context, int index) {
                        return _SwipeSelector(index);
                      },
                    ),
                  ),
                ])));
  }
}


lib/FluidActionCard/Fluid_action_card.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

class FluidActionCard extends StatefulWidget {
  //for setting the card's first gradient color
  final Color color1;
  //for setting the card's second gradient color
  final Color color2;
  //for setting the screen(container) background color
  final Color backgroundcolor;
  //for setting the card's borderradius and shape
  final BorderRadius borderRadius1;
  //for setting the card's borderradius and shape
  final BorderRadius borderRadius2;
  //for setting the card's height
  final double height;
  //for setting the card's width
  final double width;
  //for the number of cards the user wants
  final int CardCount;
  //for setting the position of the card
  final double Position;
  //for setting the shadow of the card
  final BoxShadow? shadow;
  //for gesture detection of the card
  final Function? ontap;
  //for adding text to the card
  final Text? text1;
  //for adding text to the card
  final Text? text2;
  //for setting text position on the card
  final double? TextPosition_Top;
  //for setting text position on the card
  final double? TextPosition_Down;
  //for setting image url
  final String assetimage;

  const FluidActionCard(
      {Key? key,
      required this.color1,
      required this.color2,
      required this.borderRadius1,
      required this.borderRadius2,
      required this.height,
      required this.width,
      required this.backgroundcolor,
      required this.CardCount,
      required this.Position,
      this.shadow,
      this.ontap,
      this.text1,
      this.text2,
      this.TextPosition_Top,
      this.TextPosition_Down,
      required this.assetimage})
      : super(key: key);

  @override
  _FluidActionCardState createState() => _FluidActionCardState();
}

class _FluidActionCardState extends State<FluidActionCard>
    with SingleTickerProviderStateMixin {
  TabController? _tabController;
  PageController? _pageController;
  int _selectedPage = 0;

  @override
  void initState() {
    super.initState();
    _tabController = TabController(initialIndex: 0, length: 5, vsync: this);
    _pageController = PageController(initialPage: 0, viewportFraction: 0.8);
  }

  _ActionSelector(int index) {
    return AnimatedBuilder(
      animation: _pageController!,
      builder: (BuildContext context, Widget? widget) {
        double value = 1;
        if (_pageController!.position.haveDimensions) {
          value = (_pageController!.page! - index);
          value = (1 - (value.abs() * 0.4)).clamp(0.0, 1.0);
        }
        return Center(
          child: SizedBox(
            height: Curves.easeInOut.transform(value) * 500.0,
            width: Curves.easeInOut.transform(value) * 430.0,
            child: widget,
          ),
        );
      },
      child: GestureDetector(
        onTap: () {
          widget.ontap;
        },
        child: Stack(
          alignment: Alignment.center,
          children: <Widget>[
            ClipRRect(
              borderRadius: widget.borderRadius1,
              child: Container(
                decoration: BoxDecoration(
                  gradient: LinearGradient(
                      begin: Alignment.topCenter,
                      end: Alignment.bottomCenter,
                      colors: [
                        widget.color1,
                        widget.color2,
                      ]),
                  boxShadow: [widget.shadow!],
                  borderRadius: widget.borderRadius2,
                ),
                margin: EdgeInsets.fromLTRB(10.0, 10.0, 10.0, 30.0),
                child: Stack(
                  children: <Widget>[
                    Center(
                      child: Center(
                        child: Image(
                          height: 270.0,
                          width: 290.0,
                          image: AssetImage(
                            widget.assetimage,
                          ),
                          fit: BoxFit.cover,
                        ),
                      ),
                    ),
                    Positioned(
                      top: widget.TextPosition_Top,
                      right: widget.TextPosition_Down,
                      child: Column(
                        crossAxisAlignment: CrossAxisAlignment.center,
                        children: <Widget>[
                          if (widget.text1 != null) widget.text1!,
                          if (widget.text2 != null) widget.text2!,
                        ],
                      ),
                    ),
                  ],
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        backgroundColor: widget.backgroundcolor,
        body: AnnotatedRegion<SystemUiOverlayStyle>(
            value: SystemUiOverlayStyle.dark,
            child: ListView(
                padding: EdgeInsets.symmetric(vertical: widget.Position),
                children: <Widget>[
                  Container(
                    height: widget.height,
                    width: double.infinity,
                    child: PageView.builder(
                      controller: _pageController,
                      onPageChanged: (int index) {
                        setState(() {
                          _selectedPage = index;
                        });
                      },
                      itemCount: widget.CardCount,
                      itemBuilder: (BuildContext context, int index) {
                        return _ActionSelector(index);
                      },
                    ),
                  ),
                ])));
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:Fluid_action_card/lib/FluidActionCard/fluid_action_card';


void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        title: 'Fluid Action card display',
        debugShowCheckedModeBanner: false,
        home: Scaffold(
          body: FluidActionCard(
            color1: Colors.greenAccent.shade400,
            color2: Colors.black12,
            backgroundcolor: Colors.grey[900]!,
            borderRadius1: BorderRadius.circular(20),
            borderRadius2: BorderRadius.circular(20),
            TextPosition_Top: 22.0,
            TextPosition_Down: 29.0,
            height: 400.0,
            width: 240.0,
            CardCount: 3,
            text1: Text(
              "Starts From",
              style: TextStyle(
                  color: Colors.white,
                  fontSize: 18.0,
                  fontStyle: FontStyle.italic),
            ),
            text2: Text(
              "300",
              style: TextStyle(
                color: Colors.white,
                fontSize: 20.0,
                fontWeight: FontWeight.w700,
                letterSpacing: 2.0,
              ),
            ),
            Position: 250.0,
            shadow: BoxShadow(
              color: Colors.black12,
              blurRadius: 10.0,
              spreadRadius: 0.2,
              offset: Offset(0, 3),
            ),
            assetimage: 'assets/images/plant1.png',
            ontap: () {},
          ),
        ));
  }
}
